### PR TITLE
processHTML being called too late

### PR DESCRIPTION
**Pull request by chrisleversuch**
We've noticed that in the last few days (we think since Chrome 71 has started rolling out to Android devices), the processHTML method is being called too late so it's running against the HTML from the request to www.google.com rather than the HTML form that it should be running against. The result is that onAuthorizationCompleted is called with empty strings for md and pares.

This PR uses onPageFinished to capture the HTML from each request (except the one to www.google.com) so that the reqex can be run against the correct HTML.

### CODE REVIEW

**Comment by sandyscoffable** [comment]
Thank you so much for this ❤️

**Comment by sandyscoffable** [comment]
# Replicating issue on master version

Testing using release 1.1.2:
* Webviews using Chrome 70 work fine
* Webviews using Chrome 71 don't work

👍 Seems like this is the problem.

# Testing this branch

* Webviews using Chrome 70 **don't** seem to work (blank md / pares again - the exact same issue FWICS) 😱 
* Webviews using Chrome 71 work fine 👍 

@chrisleversuch  - Can you replicate these same problems on Chrome 70 using this branch?

# Test environment

My Samsung Android phone has Chrome 70 installed by default. To get version 71 I had to install Chrome Beta.

The choice of webview to use is buried in the Developer Options on Android. 



I tested using our commercial app (development version) and used the test platform provided by our PSP.

**Comment by chrisleversuch** [comment]
Yeah I'm getting the same problem, not sure how I didn't see it earlier. I'd guess onPageStarted for  the Google request is being called before  onPageFinished for the previous request. I'll see what I can do

**Comment by chrisleversuch** [comment]
FYI I think I'll try moving the processing to onPageFinished - as soon as there's an html page with MD and PaRes inputs, trigger the completion handler. That's the html we're interested in and I don't think there's any reason to wait for the form to submit itself

**Comment by sandyscoffable** [comment]
Cool. I'm now in a bit of a quandary as to what to do. Ideally I'd like to have this fixed this week, with a gradual app rollout over the next 7 days, in time for the Christmas period (I _really_ don't want to be deploying new versions of apps while having Christmas dinner etc). 

You are much more familiar with this code than I am and have already made good progress to solving the issue, however I've got a work related deadline. So, if this isn't a high priority for you / you stop work on it, then give me a shout and I'll have a crack at it (no guarantees that I'll actually manage to fix it though). Conversely, if you do manage to fabricate a fix, give me a shout and I'll give it another test to validate it, after which we can try to get the attention of the repository maintainer to deploy a new release (as otherwise this library is dead in the water imminently).

**Comment by sandyscoffable** [comment]
> FYI I think I'll try moving the processing to onPageFinished - as soon as there's an html page with MD and PaRes inputs, trigger the completion handler. That's the html we're interested in and I don't think there's any reason to wait for the form to submit itself

I've been looking at how this library works under the hood. What you propose makes sense, however it seems a pretty fairly fundamental change to this library. I'm worried about breaking the spaghetti code around "stacked mode" which I'm not using (and I've no idea what that is - There is only a cryptic comment to go on and the commit history for those lines aren't very useful). I found https://github.com/LivotovLabs/3DSView/commit/71de5aa3b60136ce2a3a8ff359e851331cbcb0c2 but I still have little idea _why_ this was introduced.

I'm pretty sure this library could be a lot simpler than it currently is at the moment. I guess I/we could put together a fix in the manner that you suggest and ditch stacked mode, and perhaps someone else who is using that can figure out how to fix it, if it's important to them? 

FYI I'm merely using the basic `D3SView` - no frills.

**Comment by chrisleversuch** [comment]
I've updated the code with another attempt. I've tried it on Chrome 70 and I think Chrome 71 (I installed the beta and changed the WebView setting in Developer Options).
I have the same concern as you with stacked mode, I've tried to keep the changes as minimal as possible. It now runs the processing in both onPageStarted and onPageFinished, using a boolean to track when it's been done so the completion handler is only called once.

File: 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java [hunk 1]
```diff
@@ -260,11 +261,14 @@ private void completeAuthorization(String html)
...
+        if (TextUtils.isEmpty(md) || TextUtils.isEmpty(pares)) {
+            return;
+        }
+
+        postbackHandled = true;
         if (authorizationListener != null)
         {
-            // reset these flags for next time.
-            urlReturned = false;
-            postbackHandled = false;
```
[the remark this answers is not in the document]
Comment (chrisleversuch) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Yeah I considered that however they get reset every time you call authorize() so I think re-use still works

[the remark this answers is not in the document]
Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: 👍

**Comment by sandyscoffable** [comment]
Right, I've tested this 10 times on both Chrome 70 and 71 and it works. Good job. 👍 

The only niggle I have is a 4-5 second delay before I can enter information on the ACS form (The ACS form is shown, for ages, I can see it, but I can't press on the text entry fields for what feels like *ages*). The issue seems to be on both Chrome 70 and 71.

I rolled back to 1.1.2, and that seems rapid in comparison on both versions of Chrome (albeit 71 fails with the issue we are trying to fix here).

**Comment by sandyscoffable** [comment]
Video of issue here: https://drive.google.com/file/d/1I8ImPRULPQAyyhwCzmhzRi-FxgUul8ez/view?usp=sharing

**Comment by chrisleversuch** [comment]
I'm not sure what would cause the delay and I don't think I'm seeing the same issue - normally I just tap submit when the form loads but I can interact with the text entry and drop down (I assume the form I see is different to yours - I'm not sure who the payment provider is in this project, we're just using our client's API)

**Comment by sandyscoffable** [comment]
I'll provide a video of version 1.1.2, just give me a minute

**Comment by sandyscoffable** [comment]
I notice there is some other stuff on `master` since 1.1.2 was released.

What I might try doing is branching from 1.1.2 and cherry-picking your changes on top, to see if it's your changes, or the other stuff that's causing the issue.

**Comment by sandyscoffable** [comment]
Ok ... `master` version doesn't appear to exhibit the issue.

File: 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java [hunk 2]
```diff
@@ -157,6 +160,16 @@ public void onPageStarted(WebView view, String url, Bitmap icon)
...
             }
 
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                super.onPageFinished(view, url);
+
+                if (url.toLowerCase().contains(postbackUrl.toLowerCase())) {
+                    return;
+                }
+                view.loadUrl(String.format("javascript:window.%s.processHTML(document.getElementsByTagName('html')[0].innerHTML);", JavaScriptNS));
```
Comment (sandyscoffable) [inline 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Commenting out this line fixes the responsiveness issue.

Comment (chrisleversuch) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: I assume that re-introduces the original issue on Chrome 71?

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Yeah it totally breaks the library. I'm trying to avoid this on first page load to see if that improves things.

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Winner, it does. I'll push those changes now.

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: aaanndd ... I can't push to your branch

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]:
```diff
diff --git a/3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java b/3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java
index 94af1ab..3620bd0 100755
--- a/3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java
+++ b/3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java
@@ -72,6 +72,8 @@ public class D3SView extends WebView
      * When set to <b>true</b>, SSL certificate errors and self-signed certificates errors will be ignored.
      */
     private boolean debugMode = false;
+    
+    private boolean initialPageLoadCompleted = false;
 
     /**
      * Url that will be used by ACS server for posting result data on authorization completion. We will be monitoring
@@ -167,9 +169,14 @@ public class D3SView extends WebView
                 if (url.toLowerCase().contains(postbackUrl.toLowerCase())) {
                     return;
                 }
-                view.loadUrl(String.format("javascript:window.%s.processHTML(document.getElementsByTagName('html')[0].innerHTML);", JavaScriptNS));
+                
+                if(!initialPageLoadCompleted) {
+                               initialPageLoadCompleted = true;
+                } else {
+                               view.loadUrl(String.format("javascript:window.%s.processHTML(document.getElementsByTagName('html')[0].innerHTML);", JavaScriptNS));
+                }
             }
-;
+
             public void onReceivedError(WebView view, int errorCode, String description, String failingUrl)
             {
                 if (!failingUrl.startsWith(postbackUrl))
@@ -332,6 +339,7 @@ public class D3SView extends WebView
     {
         urlReturned = false;
         postbackHandled.set(false);
+        initialPageLoadCompleted = false;
 
         if (authorizationListener != null)
         {
```

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Seems to work on both Chrome 70 and 71

**Comment by sandyscoffable** [comment]
@livotov - would you be able to review this PR? It fixes this library on the latest version of Chrome (71)

**Comment by rcpassos** [comment]
I need this fix

File: 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java [hunk 3]
```diff
@@ -248,11 +271,13 @@ private void completeAuthorization(String html)
             }
         }
 
-        if (authorizationListener != null)
+        // If we didn't find both MD and PaRes, don't continue
+        if (TextUtils.isEmpty(md) || TextUtils.isEmpty(pares)) {
```
Comment (markomeara-whitbread) [inline 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Your comment says "both MD and Pares", but your check is an OR. Is the OR correct?

Comment (chrisleversuch) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Yeah we need both so if 1 is empty, return.

Comment (sandyscoffable) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Even though I'm not the author, I'll jump in my my 2c.

Yes, I believe the OR is correct. Both pieces of information are needed to continue with the payment.

Caveat: I'm not 100% sure what I'm about to say is correct - I believe the MD is some sort of reference for the verification process that's passed around, so it's _conceivable_ the MD could be present without the PARes (the important bit that signifies the end of the process ... the Payer Authentication Response).

TL;DR: Unless I see evidence suggesting otherwise, I'd vote for keeping the code as is.

Comment (markomeara-whitbread) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Thanks, yes the OR makes sense thinking about it, just wanted to confirm

Comment (markomeara-whitbread) [reply 3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java]: Follow up - I've done a quick bit of research and apparently the MD parameter is optional according to some specifications, eg:

https://developer.paysafe.com/en/rest-apis/3ds/typical-api-calls/submit-a-request-to-the-3d-secure-acs-server-and-retrieve-a-pares/

Though other specifications I have found say it's required. So, conflicting information.

**Comment by 1240** [comment]
hey,  what are we waiting for?

**Comment by rihantvx** [comment]
Hello, @sandyscoffable have you found a solution yet? I am also experiencing the same issue with capital one cards. 

I have had a successful transaction with 454313 btw.

**Comment by sandyscoffable** [comment]
@rihantvx - 454313 is possibly a red-herring ... it's quite difficult to get a definitive list as consumers can legitimately abandon the payment process at that point.

Clutching at straws, but I wonder if the issue is with the regex we use. I've not got a card to hand that I can try (I've been embroiled in other stuff and the employee that has a co-op card has gone home). I've got a branch of this that spits out the HTML that I can inspect, I just need a card that I can test with 🚶 

Pretty close to just creating a co-op bank account just to test this TBH!

**Comment by markomeara-whitbread** [comment]
We also experienced continuing issues even after adopting this change, about 30% of payments still did not complete successfully from 3DSecure screen. We're now moving to a more elaborate solution involving back-end callbacks and no longer using this library.

**Comment by sandyscoffable** [comment]
@markomeara-whitbread - I assume you mean you have a dedicated endpoint in your server side stack that the consumer is redirected to and the `PaRes` is extracted? Makes sense I guess.. I could totally do that too..

**Comment by sandyscoffable** [comment]
I'm going to continue to try and debug / fix this issue.

**Comment by sandyscoffable** [comment]
I've been able to reproduce the issue here. When the library loads it seems to fairly quickly show the above Google page in the screenshot (without any sort of intervention required).

The code that extracts the PaRes from the HTML doesn't appear to be called (as I had debug code to spit out the HTML if that was hit). There are the following log lines though:

```
2019-01-11 12:55:07.033 10790-12869/com.scoffable W/chromium: [WARNING:spdy_session.cc(2876)] Received RST for invalid stream1
2019-01-11 12:55:07.079 10790-10790/com.scoffable I/chromium: [INFO:CONSOLE(1)] "Uncaught TypeError: Cannot read property 'innerHTML' of undefined", source: https://www.google.com/ (1)
```

So, still not sure why this is happening, but it's definitely an issue. The employee who has the affected card is now off until Monday, so I can't reproduce again until then.

**Comment by rihantvx** [comment]
@sandyscoffable Did you manage to reproduce this with Co-op card? 

I am yet to find someone in our company who has a co-op card :-(

**Comment by livotov** [comment]
Thank you guys for awesome discussion and PR!
Sorry for delayed merge, was quite hard and busy year, especially end of it.

**Comment by sandyscoffable** [comment]
@rihantvx - Sorry I've been embroiled in other high priority projects. I do have this on my list to pick up in the near future.

**Comment by sandyscoffable** [comment]
I've created https://github.com/LivotovLabs/3DSView/issues/23 to track the issue
